### ISSUE TITLE
incorrect branch reporting with __debug__

### ISSUE THREAD

**Issue by nedbat** [coveragepy/coveragepy#522]
*Originally reported by* **John Theodore Goetz (Bitbucket: [theodoregoetz](https://bitbucket.org/theodoregoetz), GitHub: [theodoregoetz](https://github.com/theodoregoetz))**

----------------------------------------

Coverage report indicates a missing branch with ``__debug__`` within an ``if``/``else`` block. Note if a ``pass`` is added after the ``if __debug__`` then coverage is correctly shown as 100%.

```
#!python

for value in [True, False]:
    if value:
        if __debug__:
            x = 1
        #pass  # required for coverage to show 100%
    else:
        x = 2
```

Running coverage in both ``__debug__`` and with ``python -O`` (appending the results):

```
#!bash

python -mcoverage erase
python -mcoverage run --branch prog.py
python -O -mcoverage run --branch --append prog.py
python -mcoverage report -m
```

Results in the following:

```
#!bash

Name       Stmts   Miss Branch BrPart  Cover   Missing
------------------------------------------------------
prog2.py       4      0      4      1    88%   2->1
```

----------------------------------------
- Bitbucket: https://bitbucket.org/ned/coveragepy/issue/522

**Comment by nedbat** [comment]
*Original comment by* **Loic Dachary (Bitbucket: [dachary](https://bitbucket.org/dachary), GitHub: [dachary](https://github.com/dachary))**

----------------------------------------

Reproduced the issue as instructed at [24aff3d7bfd5 (bb)](https://bitbucket.org/ned/coveragepy/commits/24aff3d7bfd5) with python-2.7.12 and python-3.5

**Comment by nedbat** [comment]
*Original comment by* **Loic Dachary (Bitbucket: [dachary](https://bitbucket.org/dachary), GitHub: [dachary](https://github.com/dachary))**

----------------------------------------

As of [813a3f0 (bb)](https://bitbucket.org/ned/coveragepy/commits/813a3f0) the coverage shows 100%, the bug has been fixed and this issue can be closed

**Comment by nedbat** [comment]
*Original comment by* **Loic Dachary (Bitbucket: [dachary](https://bitbucket.org/dachary), GitHub: [dachary](https://github.com/dachary))**

----------------------------------------

@nedbat is there anything I can do before this issue can be closed ?

**Comment by nedbat** [comment]
Using the latest code, I still see the behavior originally reported.  Leaving open.

**Comment by nedbat** [comment]
A simpler case will show the problem:
```
$ cat bug522.py
one = 1
for value in [True, False]:
    if value:
        if one:
            x = 1
    else:
        x = 2

for value in [True, False]:
    if value:
        if 1:
            x = 1
    else:
        x = 2
$ python -m coverage run --branch bug522.py
$ python -m coverage report -m
Name        Stmts   Miss Branch BrPart  Cover   Missing
-------------------------------------------------------
bug522.py      10      0     10      2    90%   4->2, 10->9
```
The "if 1:" statement doesn't appear in the compiled bytecode, so it's being removed from consideration, which makes the branch be attributed to 10->9 rather than 11->9.

**Comment by nedbat** [comment]
We treat constant expressions specially in while loops, but not in if's.

**Comment by nedbat** [comment]
Fixed in [7d90fa43b338 (bb)](https://bitbucket.org/ned/coveragepy/commits/7d90fa43b338).

**Comment by nedbat** [comment]
This was shipped in 4.3.2.
